### PR TITLE
Fixing flacky test_id:3203

### PR DESCRIPTION
**Pull request by Barakmor1**
Signed-off-by: bmordeha <bmodeha@redhat.com>

**What this PR does / why we need it**:
Currently the supportedFeatures var is set to be all the Features that are supported
in at least one node in the cluster but in tests we assume that this set of features exist 
in all the nodes of the cluster.
This might cause flakiness if the cluster is heterogeneous and some of the features 
are not supported in all node as this test try to avoid scheduling vm to node with one of
those features and verify that the VMI is unschedulable but it can still schedule the vm to a node
that doesn't support this feature.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #
To fix this i modifed the the test to consider only features that are supported in all the nodes.
**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by xpivarc** [comment]
Fixes https://github.com/kubevirt/kubevirt/issues/7792

File: tests/utils.go [hunk 1]
```diff
@@ -469,6 +469,64 @@ func GetSupportedCPUModels(nodes k8sv1.NodeList) []string {
 	return cpus
 }
 
+func GetSupportedCPUFeaturesAmongAllTheNodes(nodes k8sv1.NodeList) []string {
```
Comment (maiqueb) [inline tests/utils.go:472]: IMO, doesn't make sense to name this function `GetSupportedCPUFeaturesAmongAllTheNodes` and pass it a list of nodes. What if you pass it a subset of nodes ? 

Either make the func read the full list of nodes, or name the method `GetSupportedCPUFeaturesFromNodes` - or something similar. 

My preference is clearly the later option (renaming the func).

File: tests/utils.go [hunk 2]
```diff
@@ -469,6 +469,64 @@ func GetSupportedCPUModels(nodes k8sv1.NodeList) []string {
...
+	var featureDenyList = map[string]bool{
```
Comment (maiqueb) [inline tests/utils.go:473]: IMO, the type can be improved; if you're really just implementing a set, you can make it a map of `map[string]struct{}` , which is preferable, since an empty struct occupies no memory.

File: tests/utils.go [hunk 3]
```diff
@@ -469,6 +469,64 @@ func GetSupportedCPUModels(nodes k8sv1.NodeList) []string {
...
+		"svm": true,
+	}
+	var supportedFeatures []string
+	for label := range nodes.Items[0].Labels {
+		if strings.Contains(label, services.NFD_CPU_FEATURE_PREFIX) {
+			feature := strings.TrimPrefix(label, services.NFD_CPU_FEATURE_PREFIX)
+			if _, ok := featureDenyList[feature]; !ok {
+				supportedFeatures = append(supportedFeatures, feature)
+			}
+		}
```
Comment (maiqueb) [inline tests/utils.go:483]: you can export this into a func and re-use [below](https://github.com/kubevirt/kubevirt/pull/7800/files#diff-847cd06c1bf6ef597ea96894f8b32ecb4d074ad1bc785b39871e11fb6ab71b3dR489-R494) - iiuc, this code is duplicate.

File: tests/utils.go [hunk 4]
```diff
@@ -469,6 +469,64 @@ func GetSupportedCPUModels(nodes k8sv1.NodeList) []string {
...
+
+	}
+
+	for _, element := range secondSet {
+		if firstSetBoolMap[element] {
+			setOfFeaturesWithDups = append(setOfFeaturesWithDups, element)
+		}
+	}
+	intersection = removeDups(setOfFeaturesWithDups)
+	return intersection
```
Comment (maiqueb) [inline tests/utils.go:516]: nit: why not `return removeDups(setOfFeaturesWithDups)` ?

File: tests/utils.go [hunk 5]
```diff
@@ -469,6 +469,64 @@ func GetSupportedCPUModels(nodes k8sv1.NodeList) []string {
...
+}
+
+func removeDups(elements []string) (intersection []string) {
+	found := make(map[string]bool)
+	for _, element := range elements {
+		if found[element] == false {
```
Comment (maiqueb) [inline tests/utils.go:522]: I'm trying to understand how can this ever happen ... afaiu, you never `map[key] = false` ... While I'm aware the default for a boolean is false, how can you add a key to a map without stating its value ? 

Furthermore, I think `if _, foundItem := found[element]; !foundItem { ... }` to be more golang idiomatic.

File: tests/utils.go [hunk 6]
```diff
@@ -469,6 +469,64 @@ func GetSupportedCPUModels(nodes k8sv1.NodeList) []string {
...
+	for _, element := range secondSet {
+		if firstSetBoolMap[element] {
+			setOfFeaturesWithDups = append(setOfFeaturesWithDups, element)
+		}
+	}
+	intersection = removeDups(setOfFeaturesWithDups)
+	return intersection
+}
+
+func removeDups(elements []string) (intersection []string) {
```
Comment (maiqueb) [inline tests/utils.go:519]: Named return values are imo very hard to follow.

Can't you just return the slice ?

**Review by maiqueb — CHANGES_REQUESTED** [review]
I think you shouldn't add this code to `tests/utils.go` - we're actually trying to slim that file down.

Since this code is only being used in one test suite, would you define this func there ?

File: tests/utils.go (hunk 1 above)
Comment (Barakmor1) [reply tests/utils.go]: Hey Thanks for reviewing.
Done.

File: tests/utils.go (hunk 2 above)
Comment (Barakmor1) [reply tests/utils.go]: Oh i didn't know this trick nice!
Done.

File: tests/utils.go (hunk 3 above)
Comment (Barakmor1) [reply tests/utils.go]: Done.

File: tests/utils.go (hunk 4 above)
Comment (Barakmor1) [reply tests/utils.go]: Done

File: tests/utils.go (hunk 5 above)
Comment (Barakmor1) [reply tests/utils.go]: Done

File: tests/utils.go (hunk 6 above)
Comment (Barakmor1) [reply tests/utils.go]: i see you point but this function is very smalland i don't see much difference between declaring the return var in the func's signature or in the first line of the func. :)

File: tests/vmi_lifecycle_test.go [hunk 7]
```diff
@@ -1186,17 +1186,74 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
...
+						if _, exist := firstSetMap[element]; exist {
+							setOfFeaturesWithDups = append(setOfFeaturesWithDups, element)
+						}
+					}
+					return removeDups(setOfFeaturesWithDups)
+				}
+
+				GetSupportedCPUFeaturesFromNodes := func(nodes k8sv1.NodeList) []string {
+					var supportedFeatures []string
+					for label := range nodes.Items[0].Labels {
```
Comment (maiqueb) [inline tests/vmi_lifecycle_test.go:1231]: you'll need to assure `nodes.Items` is not empty, right ? (I'm aware it's probably stupid to run the test is your node list is empty ... :) )

IIUC, here you're just initializing the list of supported features. Would it make sense to generalize this into a single for ?

**Review by maiqueb — COMMENTED** [review]
Looks good, nice job.

I just have a question.

File: tests/utils.go (hunk 6 above)
Comment (maiqueb) [reply tests/utils.go]: fair enough.

File: tests/vmi_lifecycle_test.go (hunk 7 above)
Comment (Barakmor1) [reply tests/vmi_lifecycle_test.go:1231]: Hey ,
I will check that the nodes list isn't empty np although the tests will fail much earlier in that case ,i think that they won't run at all.

In that function i perform sets intersection of many sets and if i'll initial the first set to be empty the result will be empty 
because intersection with empty set is empty set to i initialed it to the first set of labels.

I can do it with boolean but i think that it will be much more complicated and less understandable

Comment (Barakmor1) [reply tests/vmi_lifecycle_test.go:1231]: Honestly i can't think on A clearer alternative but if anyone have idea i'll be happy to hear.
Also don't forget this is an anonymous function  that run only in one test and we usually 
run the tests with not more than 10 nodes with less than 200 labels so the performance 
shouldn't be a problem in that case.

**Review by maiqueb — APPROVED** [review]
Thanks for addressing this flake. 

Do you think this approach could be re-used to mitigating [[test_id:1639]the vmi with cpu.model matching a nfd label on a node should be scheduled](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/7800/pull-kubevirt-e2e-k8s-1.21-sig-compute/1529782638850609152) (I think) similar flake ?

**Comment by Barakmor1** [comment]
> Thanks for addressing this flake.
> 
> Do you think this approach could be re-used to mitigating [[test_id:1639]the vmi with cpu.model matching a nfd label on a node should be scheduled](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/7800/pull-kubevirt-e2e-k8s-1.21-sig-compute/1529782638850609152) (I think) similar flake ?

I think that this isn't a flacky test ,it's a bug and it was handled here:
https://github.com/kubevirt/kubevirt/pull/7770

**Comment by Barakmor1** [comment]
@xpivarc Hey can you have a look please?

**Comment by vladikr** [comment]
/approve
Thanks Barak

File: tests/vmi_lifecycle_test.go [hunk 8]
```diff
@@ -1186,17 +1186,74 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
...
+
+					for _, element := range secondSet {
+						if _, exist := firstSetMap[element]; exist {
+							setOfFeaturesWithDups = append(setOfFeaturesWithDups, element)
+						}
+					}
+					return removeDups(setOfFeaturesWithDups)
+				}
+
+				GetSupportedCPUFeaturesFromNodes := func(nodes k8sv1.NodeList) []string {
```
Comment (xpivarc) [inline tests/vmi_lifecycle_test.go:1229]: `GetSupportedCPUFeaturesFromNodes`` -> `getSupportedCPUFeaturesFromNodes`

File: tests/vmi_lifecycle_test.go [hunk 9]
```diff
@@ -1186,17 +1186,74 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
...
+							intersection = append(intersection, element)
+							found[element] = struct{}{}
+						}
+					}
+					return intersection
+				}
+
+				setIntersection := func(firstSet, secondSet []string) []string {
+					firstSetMap := make(map[string]struct{})
+					var setOfFeaturesWithDups []string
```
Comment (xpivarc) [inline tests/vmi_lifecycle_test.go:1215]: Why don't use a map? You would save one function.

File: tests/vmi_lifecycle_test.go [hunk 10]
```diff
@@ -1186,17 +1186,74 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
...
+					for _, element := range elements {
+						if _, exist := found[element]; !exist {
+							intersection = append(intersection, element)
+							found[element] = struct{}{}
+						}
+					}
+					return intersection
+				}
+
+				setIntersection := func(firstSet, secondSet []string) []string {
```
Comment (xpivarc) [inline tests/vmi_lifecycle_test.go:1213]: I think `intersection` is clear enough

**Review by xpivarc — COMMENTED** [review]
I guess something to consider is large clusters. We could simply take 4 nodes into consideration. WDYT?

**Comment by maiqueb** [comment]
Another thing I just remembered is we could use the k8s [sets package](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/sets) instead of re-implementing them. We're already using the apimachinery dependency all over the place ...

For what it's worth, they do what @xpivarc is suggesting on https://github.com/kubevirt/kubevirt/pull/7800#discussion_r898433981 (name that function `Intersection`).
